### PR TITLE
Return links to IRM results from the query endpoint

### PR DESCRIPTION
**Pull request by figliold**
Return links to IRM results from the query endpoint

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -120,17 +120,17 @@ After configuring the build system:
 
 ## Using
 ### Querying
-The Search service can be queried for Products. See [Query Service API](https://github.com/connexta/ion-query-api) for information about paths and parameters
+The Search service can be queried for IRM. See [Query Service API](https://github.com/connexta/ion-query-api) for information about paths and parameters
```
Comment (ahoffer) [inline README.md]: Food for thought. 
From a "RESTful" perspective, the Search service is querying for a resource. In this case the resource is "dataset". 
The most RESTful thing to do is to return a URI to the dataset itself. 

In a RESTful paradigm, IRM is an XML document that is one "representation" of a dataset. To get an IRM document, the client would send a GET to the dataset's URI with an accept header. The message would look like:

> GET /dataset
> HOST store.com
> ACCEPT: irm/xml

The media type is `irm/xml`. I believe that applications should come up with media types meaningful their applications. It makes a more complete model of the domain. 

Another alternative would be to use a query param,  but it is "RESTful"

`http://host/dataset?representation=irm`

File: src/main/java/com/connexta/search/query/QueryManagerImpl.java [hunk 2]
```diff
@@ -104,24 +105,26 @@ static Filter getFilter(final String queryString)
...
    */
-  private List<URI> getProductUris(final List<String> matchingIds) {
+  private List<URI> getIrmUris(final List<String> matchingIds) {
     return matchingIds.stream()
         .map(
             id -> {
               try {
-                return new URI(productRetrieveEndpoint + id);
+                final String irmRetrieveEndpoint =
+                    String.format("%s/%s/%s", datasetRetrieveEndpoint, id, IRM_PATH_SEGMENT);
```
Comment (ahoffer) [inline src/main/java/com/connexta/search/query/QueryManagerImpl.java]: It is safer to use something like the [UriComponentBuilder](https://www.baeldung.com/spring-uricomponentsbuilder) to construct URIs than to use template strings (or string concatenation). Classes like UriComponentBuilder and Java's URI understand the domain of URIs. They help ensure that only valid URIs are constructed, and that characters and strings are properly escaped.

File: README.md (hunk 1 above)
Comment (emmberk) [reply README.md]: We cannot modify the API right now because it would break the thread. We have a ticket in our next sprint to re-write the Store API. I will copy your comment there.

File: README.md [hunk 3]
```diff
@@ -61,7 +61,7 @@ For more information about spotless checks see
     Example configs/search_config.yml:
     ```yaml
     endpointUrl:
-      productRetrieve: http://localhost:9041/mis/product/
+      datasetRetrieve: http://localhost:9041/dataset
```
Comment (emmberk) [inline README.md]: ❗️Update the rest of the repo to use the right terms. E.g.
- `product` should never be used anywhere in the repo
- `datasetId` instead of `id`

File: build.gradle [hunk 4]
```diff
@@ -8,6 +8,7 @@
 buildscript {
     ext {
         indexApiVersion = "0.2.0"
+        queryApiVersion = "0.3.0"
```
Comment (emmberk) [inline build.gradle]: ✏️ Right now, we only extract version properties that are used in more than one place. Since this is only used once, it can be inlined.

File: README.md [hunk 5]
```diff
@@ -120,17 +120,17 @@ After configuring the build system:
...
+The Search service can be queried for IRM. See [Query Service API](https://github.com/connexta/ion-query-api) for information about paths and parameters
 The Search service supports the OGC Catalogue Common Query Language (OGC CommonQL).
 See [Annex B - BNF Definition of OGC CommonQL](http://docs.opengeospatial.org/is/12-168r6/12-168r6.html#62) for the definition of the grammar.
 
 The following are valid query attributes:
 
 | Attribute | Description  |
 |---|---|
-| `id` | The unique ID that is associated with each Product |
+| `id` | The unique ID that is associated with each IRM |
```
Comment (emmberk) [inline README.md]: ❗️`dataset`

File: README.md [hunk 6]
```diff
@@ -120,17 +120,17 @@ After configuring the build system:
...
 | `contents` | A client can perform keyword queries with this attribute. |
-| `media_type` | The media (MIME) type of the Product |
+| `media_type` | The media (MIME) type of the IRM |
```
Comment (emmberk) [inline README.md]: ❗️`file`

File: README.md (hunk 1 above)
Comment (emmberk) [inline README.md]: ✏️ Missing a period at the end of this sentence.

Comment (emmberk) [inline README.md]: ✏️ `datasets` instead of `IRM`

File: src/main/java/com/connexta/search/query/QueryManagerImpl.java [hunk 7]
```diff
@@ -36,12 +36,13 @@
...
+  @NotBlank private final String datasetRetrieveEndpoint;
   @NotNull private final DataStore dataStore;
 
   public QueryManagerImpl(
-      @NotNull final DataStore dataStore, @NotBlank final String productRetrieveEndpoint) {
-    this.productRetrieveEndpoint = productRetrieveEndpoint;
+      @NotNull final DataStore dataStore, @NotBlank final String datasetRetrieveEndpoint) {
+    this.datasetRetrieveEndpoint = datasetRetrieveEndpoint;
     this.dataStore = dataStore;
   }
```
Comment (emmberk) [inline src/main/java/com/connexta/search/query/QueryManagerImpl.java]: ✏️ `@AllArgsConstructor`

File: src/test/java/com/connexta/search/common/SearchITests.java [hunk 8]
```diff
@@ -206,19 +214,22 @@ public void testStoreMetadataCstWhenSolrIsNotEmpty() throws Exception {
...
     assertThat(
         (List<String>) restTemplate.getForObject(queryUriBuilder.build(), List.class),
-        hasItem(equalTo(productLocation)));
+        hasItem(equalTo(irmLocation)));
   }
 
   @Test
-  @Disabled("TODO check that the product exists before storing cst")
-  public void testStoreMetadataProductIdNotFound() {}
+  @Disabled("TODO check that the IRM exists before storing cst")
```
Comment (emmberk) [inline src/test/java/com/connexta/search/common/SearchITests.java]: ❗️`@Disabled("TODO check that the dataset exists before indexing")`

File: src/test/java/com/connexta/search/common/SearchITests.java [hunk 9]
```diff
@@ -206,19 +214,22 @@ public void testStoreMetadataCstWhenSolrIsNotEmpty() throws Exception {
...
+  public void testStoreMetadataIrmIdNotFound() {}
```
Comment (emmberk) [inline src/test/java/com/connexta/search/common/SearchITests.java]: ❗️`datasetId`

File: src/test/java/com/connexta/search/common/SearchITests.java [hunk 10]
```diff
@@ -206,19 +214,22 @@ public void testStoreMetadataCstWhenSolrIsNotEmpty() throws Exception {
...
 
   @Test
-  public void testStoreWhenProductHasAlreadyBeenIndexed() throws Exception {
-    // given index a product
+  public void testStoreWhenIrmHasAlreadyBeenIndexed() throws Exception {
```
Comment (emmberk) [inline src/test/java/com/connexta/search/common/SearchITests.java]: ❗️ `testStoreWhenDatasetHasAlreadyBeenIndexed`

File: README.md (hunk 6 above)
Comment (emmberk) [reply README.md]: Actually, this _is_ currently the mediatype of the file attachment, which is the IRM. I will change this back. 🤦‍♀

File: build.gradle [hunk 11]
```diff
@@ -41,9 +41,10 @@ dependencies {
     annotationProcessor "org.projectlombok:lombok:${lombokVersion}"
     compileOnly "org.projectlombok:lombok:${lombokVersion}"
     implementation "com.connexta.search:index-api-rest-spring-stubs:${indexApiVersion}"
-    implementation "com.connexta.search:query-api-rest-spring-stubs:0.2.0"
+    implementation "com.connexta.search:query-api-rest-spring-stubs:0.3.0"
     implementation "io.springfox:springfox-swagger2:${springfoxSwaggerVersion}"
     implementation "io.springfox:springfox-swagger-ui:${springfoxSwaggerVersion}"
+    implementation "org.springframework:spring-web:5.2.0.RELEASE"
```
Comment (emmberk) [inline build.gradle]: ❓Why was this added?

Comment (emmberk) [reply build.gradle]: I was able to full build without this. If it is needed, it should be caught in the hero.

**Comment by emmberk** [comment]
@ahoffer is going to hero this.

**Comment by ahoffer** [comment]
I have heroed this locally.
1. I used the index endpoint to create an entry.
2. I used the query endpoint to search for the entry.
3. The endpoint returns a JSON array of of (1) URI

Here is the screen shot from Postman
image
